### PR TITLE
Prefs [TASK][28577] Create Utils class Remote Datasource

### PR DESCRIPTION
**Pull request by viet06061999**
## Related Tickets
- [Task#28577: Create Ultils Class Remote Datasource](https://edu-redmine.sun-asterisk.vn/issues/28577)

## WHAT
Create ultils class for remote datasource
## HOW
Create class AppExcutor, object HttpUltils
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

**Review by tuannm-1276 — APPROVED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/AppExcutor.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+                    NUMBER_OF_CORES * 2,
+                    KEEP_ALIVE_TIME,
+                    TimeUnit.SECONDS,
+                    LinkedBlockingQueue<Runnable>()
+                ),
+                MainThreadExecutor()
+            )
+        }
+    }
+}
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/AppExcutor.kt]: Dùng Async Task nhé

File: app/src/main/java/com/sunasterisk/travelapp/data/source/remote/utils/RemoteAsync.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class RemoteAsync(val callbackListener: OnDataCallback<String>) : AsyncTask<String, Unit, Any>() {
+    override fun doInBackground(vararg params: String?): Any =
+        HttpUtils.getApi(params[0].toString())
+
+    override fun onPostExecute(result: Any?) {
+        super.onPostExecute(result)
+        result?.let{
+            if ( it is Throwable) callbackListener.onError(it)
+            else callbackListener.onSuccess(it.toString())
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/remote/utils/RemoteAsync.kt]: 1. format code 
2. convention https://developer.android.com/kotlin/style-guide#expressions

File: app/src/main/java/com/sunasterisk/travelapp/data/source/remote/utils/RemoteAsync.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        } ?: callbackListener.onError(NullPointerException(NULL_EXCEPTION))
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/remote/utils/RemoteAsync.kt]: sao laij null pointer exception 
Exception thôi

Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/data/source/remote/utils/RemoteAsync.kt]: Vâng ạ

**Pull request merged**